modules/dragster/src/main/ts/ephox/dragster/datatransfer/DataTransfer.ts

```diff
@@ -0,0 +1,131 @@
+
+const createDataTransfer = (): DataTransfer => {
+  const dataTransferImpl = new window.DataTransfer();
+  let dropEffect: DropEffect = 'move';
+  let effectAllowed: EffectAllowed = 'all';
+
+  const dataTransfer: DataTransfer = {
+    get dropEffect() {
+      return dropEffect;
+    },
+
+    set dropEffect(effect: DataTransfer['dropEffect']) {
+      if (Arr.contains(validDropEffects, effect)) {
+        dropEffect = effect;
+      }
+    },
+
+    get effectAllowed() {
+      return effectAllowed;
+    },
+
+    set effectAllowed(allowed: DataTransfer['effectAllowed']) {
+      // TINY-9601: Only allow setting effectAllowed to a valid value in a dragstart event
+      // https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/effectAllowed
+      if (isInDragStartEvent(dataTransfer) && Arr.contains(validEffectAlloweds, allowed)) {
+        effectAllowed = allowed;
+      }
+    },
+
+    get items() {
+      return normalizeItems(dataTransfer, dataTransferImpl.items);
+    },
+
+    get files() {
+      if (isInProtectedMode(dataTransfer)) {
+        return createEmptyFileList();
+      } else {
+        return dataTransferImpl.files;
+      }
+    },
+
+    get types() {
+      return dataTransferImpl.types;
+    },
+
+    setDragImage: (image: Element, x: number, y: number): void => {
+      if (isInReadWriteMode(dataTransfer)) {
+        setDragImage(dataTransfer, { image, x, y });
+        dataTransferImpl.setDragImage(image, x, y);
+      }
+    },
+
+    getData: (format: string): string => {
+      if (isInProtectedMode(dataTransfer)) {
+        return '';
+      } else {
+        return dataTransferImpl.getData(format);
+      }
+    },
+
+    setData: (format: string, data: string): void => {
+      if (isInReadWriteMode(dataTransfer)) {
+        dataTransferImpl.setData(format, data);
+      }
+    },
+
+    clearData: (format?: string | undefined): void => {
+      if (isInReadWriteMode(dataTransfer)) {
+        dataTransferImpl.clearData(format);
+      }
+    }
+  };
+
+  setReadWriteMode(dataTransfer);
+
+  return dataTransfer;
+};
+
+const cloneDataTransfer = (original: DataTransfer): DataTransfer => {
+  // Create new DataTransfer object to ensure scope is not shared between original and clone
+  const clone = createDataTransfer();
+
+  // Store original mode and set to read-only to copy data
+  const originalMode = getMode(original);
+  setReadOnlyMode(original);
+
+  // Set clone event to dragstart to ensure effectAllowed can be set
+  setDragstartEvent(clone);
+
+  clone.dropEffect = original.dropEffect;
+  clone.effectAllowed = original.effectAllowed;
+  getDragImage(original).each((imageData) => clone.setDragImage(imageData.image, imageData.x, imageData.y));
+
+  // Copy string data
+  Arr.each(original.types, (type) => {
+    if (type !== 'Files') {
+      clone.setData(type, original.getData(type));
+    }
+  });
+
+  // Copy files
```

Comment:
Is some of these just stating the obvious?

Reply:
Agreed, removed such comments.